### PR TITLE
HSEARCH-1908 Avoid the need to dynamically override the Analyzer to fieldname mapping

### PR DESCRIPTION
**Pull request by Sanne**
- https://hibernate.atlassian.net/browse/HSEARCH-1908

This is all to avoid invoking
 IndexWriter.addDocument( document, analyzer )

as the only option in Lucene 5 is

 IndexWriter.addDocument( document )

### CODE REVIEW

File: engine/src/main/java/org/hibernate/search/backend/impl/lucene/IndexWriterDriver.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import org.apache.lucene.document.Document;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.Query;
+import org.hibernate.search.backend.impl.lucene.analysis.ConcurrentlyMutableAnalyzer;
+import org.hibernate.search.util.impl.ScopedAnalyzer;
+
+/**
+ * Encapsulates various operations to be performed on a single IndexWriter.
+ * Avoid using {@link Workspace#getIndexWriter()} directly as it bypasses lifecycle
```
Comment (hferentschik) [inline engine/src/main/java/org/hibernate/search/backend/impl/lucene/IndexWriterDriver.java]: `Workspace` needs to be fully qualified to make this link work

File: engine/src/main/java/org/hibernate/search/backend/impl/lucene/IndexWriterDriver.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ * management of the IndexWriter such as reference counting, potentially leading to leaks.
+ *
+ * @author Sanne Grinovero <sanne@hibernate.org> (C) 2015 Red Hat Inc.
+ */
+public final class IndexWriterDriver {
```
Comment (hferentschik) [inline engine/src/main/java/org/hibernate/search/backend/impl/lucene/IndexWriterDriver.java]: Not sure I like the `Driver` part so much. Why not `IndexWriterDelegate`?

File: engine/src/main/java/org/hibernate/search/backend/impl/lucene/analysis/ConcurrentlyMutableAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	public boolean isCompatibleWith(ScopedAnalyzer analyzer) {
+		ScopedAnalyzer currentAnalyzer = current.get();
+		return currentAnalyzer.isCompositeOfSameInstances( analyzer );
+	}
+
+	/**
+	 * Correct concurrency control requires external locking!
+	 * @param analyzer
+	 */
+	public void deployNewAnalyzer(ScopedAnalyzer analyzer) {
```
Comment (hferentschik) [inline engine/src/main/java/org/hibernate/search/backend/impl/lucene/analysis/ConcurrentlyMutableAnalyzer.java]: `deploy`? Is it not just a simple `set(Scoped)Analyzer`?

Comment (hferentschik) [inline engine/src/main/java/org/hibernate/search/backend/impl/lucene/analysis/ConcurrentlyMutableAnalyzer.java]: or `updateAnalyzer`

**Comment by hferentschik** [comment]
Looking good to me apart from the minor naming comments I made

File: engine/src/main/java/org/hibernate/search/backend/impl/lucene/IndexWriterDriver.java (hunk 2 above)
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/backend/impl/lucene/IndexWriterDriver.java]: Ok, that's the same name I thought about too but I then changed it to `xDriver` as it does a bit more complex _driving_ of the IW. I'll change it to your suggestion, the Driver aspect is probably too unusual / flashy.

File: engine/src/main/java/org/hibernate/search/backend/impl/lucene/analysis/ConcurrentlyMutableAnalyzer.java (hunk 3 above)
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/backend/impl/lucene/analysis/ConcurrentlyMutableAnalyzer.java]: I'll use `updateAnalyzer` as in various experiments this was doing much more complex things than just the _set_ it turned into.. that might still need to change, I wouldn't want to make a promise about it being just a set

File: engine/src/main/java/org/hibernate/search/backend/impl/lucene/IndexWriterDriver.java (hunk 1 above)
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/backend/impl/lucene/IndexWriterDriver.java]: +1

**Comment by Sanne** [comment]
Remarks applied; it took a while to rename the core class in the second commit, still have to figure out a clever way to do that. Maybe I should sed the patch files ;)

**Comment by hferentschik** [comment]
Merged.

**Pull request merged**